### PR TITLE
Feature/channel summation

### PR DESCRIPTION
**Pull request by psurukuc**
Not fully ready to merge. Need to confirm comments and such.

### CODE REVIEW

**Comment by psurukuc** [comment]
Andrew has been implementing his time-dependent grid in Katydid and that depends on some of the updates on this branch. Plan to merge this branch to develop and then push those changes into feature/time-domain-ch-sum branch before requesting a PR.

File: Source/SpectrumAnalysis/KTChannelAggregator.cc [hunk 1]
```diff
@@ -17,15 +19,22 @@ namespace Katydid
...
-            fNGrid(30),
-            fWavelength(0.0115),
-            fIsGridDefined(false),
-	    fUseAntiSpiralPhaseShifts(false),
-	    fAntiSpiralPhaseShifts()
+        KTProcessor(name),
+        fSummedFrequencyData("agg-fft", this),
+        fPhaseChFrequencySumSlot("fft", this, &KTChannelAggregator::SumChannelVoltageWithPhase, &fSummedFrequencyData),
+        fAxialSumSlot("ax-agg-fft", this, &KTChannelAggregator::SumChannelVoltageWithPhase, &fSummedFrequencyData),
+        fActiveRadius(0.0516),
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.cc:26]: Values should be initialized to things that are valid values for those variables, but will generally _not_ be useful in an analysis.  This way all users are made to explicitly specify things in their configs and not rely on default values that might get changed. So, e.g. numbers of things initialized to 0; maxes and mins to 0 and 1, respectively; wavelength to 0.

File: Source/SpectrumAnalysis/KTChannelAggregator.cc [hunk 2]
```diff
@@ -75,104 +101,199 @@ namespace Katydid
...
-	    double phaseShift=0.0;
-	    if(fUseAntiSpiralPhaseShifts) 
-	    {
-	        phaseShift=i*2*KTMath::Pi()/channelCount;
-	    }
-	    std::pair<int,double> channelPhaseShift=std::make_pair(i,phaseShift);
-	    fAntiSpiralPhaseShifts.insert(channelPhaseShift);
-	}
-	    return true;
+        for(int i=0;i<channelCount;++i)
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.cc:104]: `int` --> `unsigned` and spaces between arguments

File: Source/SpectrumAnalysis/KTChannelAggregator.cc [hunk 3]
```diff
@@ -75,104 +101,199 @@ namespace Katydid
...
+            }
+            std::pair<int,double> channelPhaseShift=std::make_pair(i,phaseShift);
+            fAntiSpiralPhaseShifts.insert(channelPhaseShift);
+        }
+        return true;
     }
 
     bool KTChannelAggregator::SumChannelVoltageWithPhase(KTFrequencySpectrumDataFFTW& fftwData)
+    {
+        KTAggregatedFrequencySpectrumDataFFTW& newAggFreqData = fftwData.Of< KTAggregatedFrequencySpectrumDataFFTW >().SetNComponents(0);
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.cc:119]: `int` --> `unsigned` and spaces between arguments

File: Source/SpectrumAnalysis/KTChannelAggregator.cc [hunk 4]
```diff
@@ -75,104 +101,199 @@ namespace Katydid
...
+    }
+
+    bool KTChannelAggregator::PerformPhaseSummation(KTFrequencySpectrumDataFFTWCore& fftwData,KTAggregatedFrequencySpectrumDataFFTW &newAggFreqData)
     {
         const KTFrequencySpectrumFFTW* freqSpectrum = fftwData.GetSpectrumFFTW(0);
         int nTimeBins = freqSpectrum->GetNTimeBins();
         // Get the number of frequency bins from the first component of fftwData
         int nFreqBins = freqSpectrum->GetNFrequencyBins();
-        int nComponents = fftwData.GetNComponents(); // Get number of components
+        int nTotalComponents = fftwData.GetNComponents(); // Get number of components
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.cc:207]: `int` --> `unsigned` and more spaces.
Either `if (true)` or `if( true )`, but be consistent.

File: Source/SpectrumAnalysis/KTChannelAggregator.hh [hunk 5]
```diff
@@ -71,21 +78,53 @@ namespace Katydid
...
+            MEMBERVARIABLE(bool, IsUserDefinedGrid);
 
-	    //AN electron undergoiing cyclotron motion has a spiral motion and not all receving channels are in phase.
-	    //If selected this option will make sure that there is a relative phase-shift applied 
-	    MEMBERVARIABLE(bool,UseAntiSpiralPhaseShifts);
+            // A boolean to run channel aggreagator when only a partial ring is defined
+            MEMBERVARIABLE(bool, IsPartialRing);
+
+            // If IsPartialRing is true, the total number of channels is the product of the number of channels extracted from the egg file and the PartialRingMultiplicity
+            MEMBERVARIABLE(int, PartialRingMultiplicity)
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.hh:89]: unsigned?

File: Source/SpectrumAnalysis/KTChannelAggregator.hh [hunk 6]
```diff
@@ -71,21 +78,53 @@ namespace Katydid
...
+            MEMBERVARIABLE(std::string, UserDefinedGridFile);
+
+    	    //The minimum frequency value above which the channel aggregated spectrum is calculated
+            MEMBERVARIABLE(double, SummationMinFreq);
+
+	        //The maximum frequency value below which the channel aggregated spectrum is calculated
+            MEMBERVARIABLE(double, SummationMaxFreq);
+
+            // Number of axial rings/subarrays
+            MEMBERVARIABLE(int, NRings);
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.hh:101]: unsigned

File: Source/SpectrumAnalysis/KTChannelAggregator.hh [hunk 7]
```diff
@@ -71,21 +78,53 @@ namespace Katydid
...
+	        std::map<int,double> fAntiSpiralPhaseShifts; 
 
-	    ///map that stores antispiral phase shifts
-	    std::map<int,double> fAntiSpiralPhaseShifts; 
+            /// Define the grid based (one per ring)
+            /*If fIsUserDefinedGrid it true, the grid is defined based on text file defined by fUserDefinedGridFile. 
+             * If fIsUserDefinedGrid is false, defines a square grid with fNGrid*fNGrid points
+             * Returns total number of grid points defined
+             */
+            int DefineGrid(KTAggregatedFrequencySpectrumDataFFTW &newAggFreqData);
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.hh:120]: return value `unsigned`?

File: Source/SpectrumAnalysis/KTChannelAggregator.hh [hunk 8]
```diff
@@ -12,6 +12,7 @@
 #include "KTData.hh"
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.hh:12]: This file has a bunch of tab characters in it.  Please convert all to 4 spaces

**Review by nsoblath — CHANGES_REQUESTED** [review]
Very nice overall.  I didn't check all of the math, or anything, but for the most part the changes looked consistent with best practices and standards, both for C++ in general and Katydid specifically.  

Is there a test program for the channel aggregators?  If not, please add one that verifies that the processors are behaving as expected.

I have some comments that apply in many places, and I've tried to include individual comments where I've noticed individual instances, but some may have been missed.

* There are a number of places where spaces are missing between function arguments (calls and definitions) or where multiple variables are defined in one line.  Please add those spaces.
* Make sure to initialize variables where they're defined to avoid the possibility that they're used uninitialized.
* Standard practice in Katydid is to use unsigned ints for counting variables (like components or bins) if there's no chance of using a negative number.  In some cases I was pretty sure unsigned variables should be used.  In some cases I wasn't sure whether or not there was a possibility of a negative value.  Please check these and use unsigned in any case where negative values aren't expected.
* There were a few hard tab characters in at least one file.

File: Examples/ConfigFiles/GridPoints.txt [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# Template for grid point file
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Examples/ConfigFiles/GridPoints.txt:1]: Hi Noah, yes, there is no compelling reason to use text file. I work on chaging this to YAML file. Just so I unserstand correctly, you want that to be done in a new feature branch right after merging this one into develop ?

[the remark this answers is not in the document]
Comment (nsoblath) [reply Examples/ConfigFiles/GridPoints.txt:1]: That's right.  As I mentioned in one of the other comments, let's rename the relevant config parameters so they don't say "text" in their names, in anticipation of the change.  But let's not hold up the merge for this particular change.

File: Source/IO/Conversions/KT2ROOT.cc [hunk 10]
```diff
@@ -477,44 +478,63 @@ namespace Katydid
...
-        //      if(fs.GetIsSquareGrid()) nGridPoints=std::sqrt(4.0*nComponents/KTMath::Pi());
-        nGridPoints = std::sqrt(nComponents);
         double fActiveRadius = aggfs.GetActiveRadius();
-        TH2D* hist = new TH2D(name.c_str(), "Frequency Spectrum Grid", nGridPoints, -fActiveRadius, fActiveRadius, nGridPoints, -fActiveRadius, fActiveRadius);
+        int nZ=aggfs.GetNAxialPositions();
+        std::vector<TGraph2D*> graphs;
+        std::vector<unsigned> pointIndex;
+        pointIndex.assign(nZ,0);
+
+        for(int iZ=0;iZ<nZ;iZ++)
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/IO/Conversions/KT2ROOT.cc:490]: Done

File: Source/IO/Conversions/KT2ROOT.cc [hunk 11]
```diff
@@ -477,44 +478,63 @@ namespace Katydid
...
+            graphs.at(iZ)->GetYaxis()->SetTitle("Y Axis (m)");
+        }
+
         for (unsigned int iComponents = 0; iComponents < nComponents; ++iComponents)
         {
-            int xBin = iComponents / nGridPoints;
-            int yBin = iComponents % nGridPoints;
-            hist->SetBinContent(xBin + 1, yBin + 1, aggfs.GetSummedGridVoltage(iComponents));
+            double xPos,yPos,zPos;
+            aggfs.GetGridPoint(iComponents,xPos,yPos,zPos);
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/IO/Conversions/KT2ROOT.cc:501]: Done

File: Source/IO/BasicROOTFileWriter/KTBasicROOTTypeWriterSpectrumAnalysis.cc [hunk 12]
```diff
@@ -808,14 +874,17 @@ namespace Katydid
...
-        conv >> histName;
-        TH2D* aggregatedGridHistogram = KT2ROOT::CreateGridHistogram(sumData, histName);
-        aggregatedGridHistogram->SetDirectory(fWriter->GetFile());
-
-        aggregatedGridHistogram->Write(); //Redundant
-        KTDEBUG(publog, "Histogram <" << histName << "> written to ROOT file");
+        conv << "graphAggGridFFTW_" << sliceNumber;
+        string graphName;
+        conv >> graphName;
+        std::vector<TGraph2D*> aggregatedGridGraphs = KT2ROOT::CreateGridGraphs(sumData,graphName);
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/IO/BasicROOTFileWriter/KTBasicROOTTypeWriterSpectrumAnalysis.cc:880]: Done

File: Source/SpectrumAnalysis/KTAxialAggregator.cc [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+    KTAxialAggregator::~KTAxialAggregator()
+    {
+    }
+
+    bool KTAxialAggregator::Configure(const scarab::param_node* node)
+    {
+        if (node != NULL)
+        {
+            fNRings= node->get_value< signed int >("n-rings", fNRings);
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTAxialAggregator.cc:35]: That is a mistake, thanks for catching.

**Comment by psurukuc** [comment]
Hi Noah, thanks for very carefully going through the commits. I think I fixed all the suggestions expect for replacing YAML file with text file.

**Comment by psurukuc** [comment]
Actually, please hold off on reviewing. I think I may have to make a few more changes. Also, I realized I haven't actually pushed all my changes to github yet.

**Comment by psurukuc** [comment]
@nsoblath, thanks for the comments. I have implemented all the changes. There is a test program called TestChannelAggregator to test the ChannelAggregator and I modified it slightly to fit the new changes better.

**Review by nsoblath — APPROVED** [review]
This all looks good.  Thanks for all of the efforts on the this PR!

**Pull request merged**